Set Docker custom metric label for cAdvisor if custom metric volume is mounted

mwielgus opened the pull request:
As agreed with @erictune, @dchen1107 and @bgrant0607  alpha support for Custom Metics in 1.2 will be enabled by the presence of a volume (configmap) mounted in a well-defined location.

PR is based on #18295 by @jszczepkowski.

mwielgus committed:
Set Docker custom metric label for cAdvisor if custom metric volume is mounted

dchen1107 commented:
cc/ @derekwaynecarr @ncdc who are interested in this feature for 1.2 release.

ncdc commented:
cc @kubernetes/rh-cluster-infra

mwielgus commented:
k8s bot test this please

derekwaynecarr commented on pkg/kubelet/custommetrics/custom_metrics.go:
```diff
@@ -0,0 +1,48 @@
+const (
+	CustomMetricsDefinitionContainerFile = "definition.json"
+
+	CustomMetricsDefinitionDir = "/etc/custom-metrics"
```
I can't easily check right now, but do we have a convention for naming special dirs?  I kind of expected to see "Kubernetes" in the path.  @pmorie - is there a convention established that says this is where kube looks for special things or is it just /etc/something   ?

bgrant0607 replied:
cc @mikedanese

mikedanese replied:
I think a good default is /etc/kubernetes (e.g. /etc/kubernetes/manifests) but this is likely something that users will want to configure. We should also allow both json and yml config. We might want to just include this as a subobject of componetconfig.KubeletConfiguration https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/types.go#L79

dchen1107 replied:
I believe the reason @mwielgus doesn't chose /etc/kubernetes/blah... because one request from @erictune is making this not kubernetes specific one.

mwielgus replied:
Yes, I don't want put CM into the default kubernetes-related config.

erictune replied:
Seems like this goes into /etc/cadvisor?

On Tue, Feb 2, 2016 at 7:17 AM, Marcin Wielgus <notifications@github.com>
wrote:

> In pkg/kubelet/custommetrics/custom_metrics.go
> <https://github.com/kubernetes/kubernetes/pull/19944#discussion_r51581225>
> :
>
> > +limitations under the License.
> > +*/
> > +
> > +// Package custommetrics contains support for instrumenting cAdvisor to gather custom metrics from pods.
> > +package custommetrics
> > +
> > +import (
> > +	"path"
> > +
> > +	"k8s.io/kubernetes/pkg/api"
> > +)
> > +
> > +const (
> > +	CustomMetricsDefinitionContainerFile = "definition.json"
> > +
> > +	CustomMetricsDefinitionDir = "/etc/custom-metrics"
>
> Yes, I don't want put CM into the default kubernetes-related config.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19944/files#r51581225>.
>

mikedanese replied:
I agree that cadvisor configuration should go in /etc/cadvisor. I don't think /etc/custom-metrics is a good place for this. It's overly generic.

derekwaynecarr commented:
Assuming every node in the cluster runs with the flag enabled similar to how we cap CPU this seems fine.  I am not aware of any requirement that would say only schedule to this set of nodes to get custom metric support.   Aside from my one question on centralizing special dirs that Kubernetes looks at to aid discoverability when debugging things, this Lgtm.

vishh commented:
I wonder if it makes sense to make cadvisor look for custom metrics for all docker containers by default? The initial check that cadvisor performs is not expensive (just a `stat`) and we could configure cadvisor in kube setup to probe custom metrics by default. This would make it unnecessary for kubelet to perform something magical.

@dchen1107 @mwielgus WDYT?

mwielgus commented:
@vishh I would rather not do that. This is an alpha feature so the exact CM configuration format may/will change. Splitting the config from the image allows to alter the format more easily and doesn't create incentive for docker image creators to attach alpha-feature-specific confs in their image.

vishh commented:
@mwielgus: I'm not suggesting changing the config. There are two separate stages required for making cadvisor collect custom metrics.
1. Metrics configuration file should be present within the container sandbox
2. A docker label with a special key `io.cadvisor.metric` and value `path_to_config_file`.

This PR is fixing the directory of the config file. Instead we can fix the path to the config file to say `/var/cadvisor/config.json`, let users specify config file via volumes, and configure cadvisor collect custom metrics by default for all docker containers that have this file. That way kubelet will not have to be involved at all, now and even in the future.

mwielgus commented:
@vishh This could work but it would involve spreading the alpha-level hack a step further to cAdvisor so it would be harder to change (cAdvisor is also used in non-kubernetes environments) when we finally agree how custom metrics should be handled in the long run.

mwielgus commented:
Moreover, as I wrote above, I would like to prevent Docker images from having CM config until we finalize all of the apis (including cAdvisor's CM format). And fixed path to the file that could be on a mounted volume or in the image itself would allow that.  

It would be a huge mess if, with 1.X release of Kubernetes people were forced to update their images just because we changed format a bit.

mwielgus commented:
@dchen1107 Ping :)

vishh commented:
I don't think we can stop people from using the current cadvisor format
since it has gone live for a few months now. If we were to introduce
additional abstractions on top of cadvisor's config, then I can see why
kubelet needs to be involved here.

On Wed, Jan 27, 2016 at 2:27 PM, Marcin Wielgus <notifications@github.com>
wrote:

> @dchen1107 <https://github.com/dchen1107> Ping :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19944#issuecomment-175892236>
> .
>

mwielgus commented:
@vishh Unfortunately, noone has a clear vision what should be the target/final design and if there should be an additional layer of abstraction. In @jszczepkowski PR (#18295) there was one, here there is no. I know i can't stop people from using this format but I would like to maintain the tentative/alpha status until we have more time to decide what's next. IMHO this is the approach that can be reverted/changed with the lowest pain. This is also, if I understood the things correctly, the approach I agreed on-site with people mentioned in the first comment.

dchen1107 commented:
Sorry for not proceed this pr earlier due to ...

The pr looks good to me overall, except I do think @vishh has a good point here on fixing the path to the config file something like /var/cadvisor/config.json. But on another side, we don't have a full picture on how cadvisor is going to be after 1.2 / 1.3. For example, running a standalone cadvisor pod for monitoring, and have minimal cadvisor library linked with kubelet, then which cadvisor has the ownership of that fixed directory? I don't know yet. 

So I don't want to block this nice alpha feature here. @mwielgus please rebase. Thanks!

erictune commented:
We can sort out the uncommon use case of 2 cadvisors later by providing
flags on cadvisor binary and kubelet binary to override the location of the
config file.

However, I still think we should make the initial default for both
standalone cadvisor binary (used without kubernetes) and
cadvisor-in-kubernetes, to be something that has cadvisor, not kubernetes
in the name.  That shows that we intend the same mechanism to work with
non-Kubernetes orchestration systems.  Is that a big change to make right
now?

On Wed, Feb 3, 2016 at 9:49 AM, Dawn Chen <notifications@github.com> wrote:

> Sorry for not proceed this pr earlier due to ...
>
> The pr looks good to me overall, except I do think @vishh
> <https://github.com/vishh> has a good point here on fixing the path to
> the config file something like /var/cadvisor/config.json. But on another
> side, we don't have a full picture on how cadvisor is going to be after 1.2
> / 1.3. For example, running a standalone cadvisor pod for monitoring, and
> have minimal cadvisor library linked with kubelet, then which cadvisor has
> the ownership of that fixed directory? I don't know yet.
>
> So I don't want to block this nice alpha feature here. @mwielgus
> <https://github.com/mwielgus> please rebase. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19944#issuecomment-179371830>
> .
>

mwielgus commented:
rebased

mwielgus commented:
annotation PR: #20089

dchen1107 commented:
LGTM

mwielgus commented:
@k8s-bot test this

mikedanese commented on pkg/apis/componentconfig/types.go:
```diff
@@ -288,6 +288,8 @@ type KubeletConfiguration struct {
 	NodeLabels map[string]string `json:"nodeLabels"`
 	// nonMasqueradeCIDR configures masquerading: traffic to IPs outside this range will use IP masquerade.
 	NonMasqueradeCIDR string `json:"nonMasqueradeCIDR"`
+	// enable gathering custom metrics.
```
this comment should start with enableCustomMetrics (json field name) because it will be used to generate documentations

mikedanese commented on cmd/kubelet/app/options/options.go:
```diff
@@ -214,4 +215,5 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.ExperimentalFlannelOverlay, "experimental-flannel-overlay", s.ExperimentalFlannelOverlay, "Experimental support for starting the kubelet with the default overlay network (flannel). Assumes flanneld is already running in client mode. [default=false]")
 	fs.DurationVar(&s.OutOfDiskTransitionFrequency.Duration, "outofdisk-transition-frequency", s.OutOfDiskTransitionFrequency.Duration, "Duration for which the kubelet has to wait before transitioning out of out-of-disk node condition status. Default: 5m0s")
 	fs.StringVar(&s.NodeIP, "node-ip", s.NodeIP, "IP address of the node. If set, kubelet will use this IP address for the node")
+	fs.BoolVar(&s.EnableCustomMetrics, "enable-custom-metrics", s.EnableCustomMetrics, "Support for gathering custom metrics.")
```
Why not make this a directory --custom-metrics-config=/path/to/custom/metrics/dir and if the string is "" don't enable it?
